controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,448 @@
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: reason + "-",
+			Namespace:    info.hfs.ObjectMeta.Namespace,
+		},
+		InvolvedObject: corev1.ObjectReference{
+			Kind:       "HostFirmwareSettings",
+			Namespace:  info.hfs.Namespace,
+			Name:       info.hfs.Name,
+			UID:        info.hfs.UID,
+			APIVersion: metal3v1alpha1.GroupVersion.String(),
+		},
+		Reason:  reason,
+		Message: message,
+		Source: corev1.EventSource{
+			Component: "metal3-hostfirmwaresettings-controller",
+		},
+		FirstTimestamp:      t,
+		LastTimestamp:       t,
+		Count:               1,
+		Type:                corev1.EventTypeNormal,
+		ReportingController: "metal3.io/hostfirmwaresettings-controller",
+	}
+
+	info.events = append(info.events, hfsEvent)
+}
+
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=hostfirmwaresettings/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=metal3.io,resources=firmwareschemas/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *HostFirmwareSettingsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", req.NamespacedName)
+	reqLogger.Info("start")
+
+	// Get the corresponding baremetalhost in this namespace, if one doesn't exist don't continue processing
+	bmh := &metal3v1alpha1.BareMetalHost{}
+	if err = r.Get(context.TODO(), req.NamespacedName, bmh); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "could not get baremetalhost")
+	}
+
+	// Fetch the HostFirmwareSettings
+	hfs := &metal3v1alpha1.HostFirmwareSettings{}
+	info := &rInfo{log: reqLogger, hfs: hfs, bmh: bmh}
+	if err = r.Get(ctx, req.NamespacedName, hfs); err != nil {
+		if k8serrors.IsNotFound(err) {
+			// A resource doesn't exist, create one if there is a baremetalhost
+			if err = r.newHostFirmwareSettings(info, req.NamespacedName); err != nil {
+
+				return ctrl.Result{}, errors.Wrap(err, "could not create firmware settings")
+			}
+		} else {
+			// Error reading the object - requeue the request.
+			return ctrl.Result{}, errors.Wrap(err, "could not load host firmware settings")
+		}
+	}
+
+	// Create a provisioner that can access Ironic API
+	provID := bmh.Status.Provisioning.ID
+	prov, err := r.ProvisionerFactory.NewProvisioner(provisioner.BuildEmptyHostData(provID), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+	info.log.Info("created a provisioner for host", "uuid", provID)
```

Comment:
This may be confusing since Provisioner is an internal entity, not something user-observable. Do we really need to log this?

Reply:
Not really since now logging uuid. Removed